### PR TITLE
refactor(core): Refactor execution post-execute promises (no-changelog)

### PR DESCRIPTION
**Pull request by netroy**
## Summary

This PR simplifies our post-executions promises code to
1. use a single promise for any node that needs to execute after an execution finishes or fails
2. simplify removal of executions from `ActiveExecutions` to (hopefully) prevent calls to `getExecution` after the execution has already been removed.

This possible also resolves the issue in https://github.com/n8n-io/n8n/pull/10789

[PAY-1554](https://linear.app/n8n/issue/PAY-1554/error-no-active-execution-found)
[CAT-168](https://linear.app/n8n/issue/CAT-168/bug-execution-info-disappears-when-stopping-execution)

## Review / Merge checklist

- [x] PR title and summary are descriptive
- [x] Tests included

### CODE REVIEW

File: packages/cli/src/active-executions.ts [hunk 1]
```diff
@@ -125,68 +135,24 @@ export class ActiveExecutions {
...
-		execution.workflowExecution!.cancel();
-
-		// Reject all the waiting promises
-		const reason = new ExecutionCancelledError(executionId);
-		for (const promise of execution.postExecutePromises) {
-			promise.reject(reason);
-		}
-
-		this.postExecuteCleanup(executionId);
+		const execution = this.getExecution(executionId);
```
Comment (despairblue) [inline packages/cli/src/active-executions.ts:140]: This turns a lot of no-ops into errors. Which is probably a good thing, albeit annoying to users if they hit this. 🤔

File: packages/cli/src/active-executions.ts [hunk 2]
```diff
@@ -95,13 +95,23 @@ export class ActiveExecutions {
...
+			postExecutePromise,
 			status: executionStatus,
 		};
 
+		// Automatically remove execution once the postExecutePromise settles
+		void postExecutePromise.promise.finally(() => {
+			this.concurrencyControl.release({ mode: executionData.executionMode });
+			setImmediate(() => {
+				delete this.activeExecutions[executionId];
+			});
```
Comment (despairblue) [inline packages/cli/src/active-executions.ts:112]: Why wait for the next tick?

File: packages/cli/src/workflow-runner.ts [hunk 3]
```diff
@@ -490,45 +489,27 @@ export class WorkflowRunner {
...
-
-				if (executionHasPostExecutionPromises) {
-					this.logger.debug(
-						`Reading execution data for execution ${executionId} from db for PostExecutionPromise.`,
-					);
-				} else {
-					this.logger.debug(
-						`Skipping execution data for execution ${executionId} since there are no PostExecutionPromise.`,
-					);
-				}
```
Comment (despairblue) [inline packages/cli/src/workflow-runner.ts:505]: Did the case ever occur that there is no post execute promise? Just wondering about the performance implication of this.

File: packages/cli/src/active-executions.ts [hunk 4]
```diff
@@ -125,68 +138,24 @@ export class ActiveExecutions {
...
-		}
-
-		// Remove from the list of active executions
-		delete this.activeExecutions[executionId];
-
-		this.concurrencyControl.release({ mode: execution.executionData.executionMode });
+	/** Mark an execution as completed */
+	finishExecution(executionId: string, fullRunData?: IRun) {
+		const execution = this.getExecution(executionId);
+		execution.postExecutePromise.resolve(fullRunData);
```
Comment (ivov) [inline packages/cli/src/active-executions.ts:151]:
```suggestion
	/** Resolve the post-execution promise in an execution. */
	settleExecution(executionId: string, fullRunData?: IRun) {		
    this.getExecution(executionId).postExecutePromise.resolve(fullRunData);
```

File: packages/cli/src/interfaces.ts [hunk 5]
```diff
@@ -193,7 +193,7 @@ export interface IExecutionsCurrentSummary {
 export interface IExecutingWorkflowData {
 	executionData: IWorkflowExecutionDataProcess;
 	startedAt: Date;
-	postExecutePromises: Array<IDeferredPromise<IRun | undefined>>;
+	postExecutePromise: IDeferredPromise<IRun | undefined>;
```
Comment (ivov) [inline packages/cli/src/interfaces.ts:196]: Channeling @tomi here - can we add a docline explaining what this is?

File: packages/cli/src/workflow-execute-additional-data.ts [hunk 6]
```diff
@@ -880,7 +880,7 @@ async function executeWorkflow(
 		}
 
 		// remove execution from active executions
-		activeExecutions.remove(executionId, fullRunData);
+		activeExecutions.finishExecution(executionId, fullRunData);
```
Comment (ivov) [inline packages/cli/src/workflow-execute-additional-data.ts:883]: The comment above is incomplete? Not sure this adds much - I'd rather the reader find the `finally` clause.

File: packages/cli/src/active-executions.ts
Comment (ivov) [inline packages/cli/src/active-executions.ts:1]: This error has been a headache. Can we create a dedicated error e.g. `MissingActiveExecution` and enrich it with any data e.g. execution status that might help us to debug? Also maybe also a clearer message for the user, e.g. `Failed to find active execution to stop. This execution was inactive by the time it was cancelled.` (or something better)

```ts
throw new ApplicationError('No active execution found', { extra: { executionId } })
```

File: packages/cli/src/active-executions.ts [hunk 7]
```diff
@@ -95,13 +95,26 @@ export class ActiveExecutions {
...
 
+		// Automatically remove execution once the postExecutePromise settles
+		void postExecutePromise.promise
+			.catch((error) => {
+				// rethrow the error unless it's ExecutionCancelledError
+				if (!(error instanceof ExecutionCancelledError)) throw error;
+			})
+			.finally(() => {
+				this.concurrencyControl.release({ mode: executionData.executionMode });
+				delete this.activeExecutions[executionId];
```
Comment (ivov) [inline packages/cli/src/active-executions.ts:115]: Debug log for removal from active executions?

File: packages/cli/src/active-executions.ts [hunk 8]
```diff
@@ -125,68 +138,24 @@ export class ActiveExecutions {
...
-		execution.workflowExecution!.cancel();
-
-		// Reject all the waiting promises
-		const reason = new ExecutionCancelledError(executionId);
-		for (const promise of execution.postExecutePromises) {
-			promise.reject(reason);
-		}
-
-		this.postExecuteCleanup(executionId);
+		const execution = this.getExecution(executionId);
```
Comment (ivov) [inline packages/cli/src/active-executions.ts:143]: Maybe we can keep reporting this error to Sentry without surfacing it to the user? Arguably the user only cares about the execution not being active anymore.

File: packages/cli/src/active-executions.ts [hunk 9]
```diff
@@ -125,68 +138,24 @@ export class ActiveExecutions {
...
-			promise.resolve(fullRunData);
-		}
-
-		this.postExecuteCleanup(executionId);
-	}
-
-	/**
-	 * Forces an execution to stop
-	 */
+	/** Forces an execution to stop */
```
Comment (ivov) [inline packages/cli/src/active-executions.ts:141]:
```suggestion
	/** Cancel the execution promise and reject its post-execution promise. */
```

**Review by ivov — COMMENTED** [review]
Manual testing went well

File: packages/cli/src/active-executions.ts (hunk 4 above)
Comment (ivov) [reply packages/cli/src/active-executions.ts:151]: Also maybe mention that resolving it with no data means failure.

File: packages/cli/src/workflow-runner.ts [hunk 10]
```diff
@@ -491,45 +490,27 @@ export class WorkflowRunner {
...
-					runData.data = (fullExecutionData as IExecutionResponse).data;
-				}
+					data: fullExecutionData.data,
+				};
 
 				// NOTE: due to the optimization of not loading the execution data from the db when no post execution promises are present,
 				// the execution data in runData.data MAY not be available here.
 				// This means that any function expecting with runData has to check if the runData.data defined from this point
-				this.activeExecutions.remove(executionId, runData);
+				this.activeExecutions.finishExecution(executionId, runData);
```
[the remark this answers is not in the document]
Comment (netroy) [reply packages/cli/src/workflow-runner.ts]: deleted the comment

File: packages/cli/src/active-executions.ts (hunk 4 above)
Comment (netroy) [reply packages/cli/src/active-executions.ts]: renamed `finishExecution` to `finalizeExecution`.

Regarding the comment explaining the no data scenario: I want to explore the option of rejecting the promise with other error types, like timeout, and any execution errors, and move away from resolving the promise on an error.

File: packages/cli/src/active-executions.ts (hunk 8 above)
Comment (despairblue) [reply packages/cli/src/active-executions.ts:147]: I don't have a strong opinion on this, but I made the change. @netroy what do you think?

File: packages/cli/src/active-executions.ts [hunk 11]
```diff
@@ -222,7 +227,7 @@ export class ActiveExecutions {
 	private getExecution(executionId: string): IExecutingWorkflowData {
 		const execution = this.activeExecutions[executionId];
 		if (!execution) {
-			throw new ExecutionNotFoundError(executionId);
+			ErrorReporterProxy.error(new ExecutionNotFoundError(executionId));
```
Comment (netroy) [inline packages/cli/src/active-executions.ts:230]: removing the throw here changes the return type to `IExecutingWorkflowData | undefined`. We should either bring back the `throw`, or update the rest of the code to handle the `undefined`.

Comment (despairblue) [reply packages/cli/src/active-executions.ts:230]: 🤦🏾 can't wait for enabling `noUncheckedIndexedAccess`...

I'm also in favor of the second. Let's see how much code it affects.

**Review by despairblue — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.61.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.61.0)

**Pull request merged**